rb_link/src/server/getter.rs

```diff
@@ -0,0 +1,190 @@
+use super::{B2RMessage, B2RServer};
+use std::collections::{HashMap, VecDeque};
+use std::sync::{Arc, Mutex};
+
+/// A getter to get message from the bluesim by id
+pub struct IDGetter {
+    b2r_cache: Arc<Mutex<HashMap<u32, VecDeque<B2RMessage>>>>,
+}
+
+impl IDGetter {
+    /// Crate a new getter bind to the given server
+    pub fn new(server: &B2RServer) -> Self {
+        IDGetter {
+            b2r_cache: server.b2r_cache.clone(),
+        }
+    }
+    /// Return the earliest message from the probe with id.
+    /// This function will block until there is a message available for retrieval.
+    pub fn get(&mut self, id: u32) -> B2RMessage {
+        loop {
+            let mut b2r_cache = self.b2r_cache.lock().expect("Fail to lock b2r_cache");
+            if let Some(queue) = b2r_cache.get_mut(&id) {
+                if let Some(b2r_message) = queue.pop_front() {
+                    return b2r_message;
+                }
+            }
+            drop(b2r_cache);
+        }
+    }
+
+    /// Get all message send by the probe with id.
+    pub fn get_id_all(&mut self, id: u32) -> Vec<B2RMessage> {
```

Comment:
again, naming!
get_id_all means get all IDs, but this method returns messages instead of IDs.